<p>
Sharing some naming conventions is a key point to make it possible for a team to efficiently collaborate. 
This rule allows to check that all method names match a provided regular expression.
</p>

<h2>Noncompliant Code Example</h2>

<p>With default provided regular expression: ^[A-Z][A-Za-z0-9]{2,30}$</p>

<pre>
class MyClass {
    void my_method(int a, int b);
    ...
</pre>

<h2>Compliant Solution</h2>

<pre>
class MyClass {
    void MyMethod(int a, int b);
    ...
</pre>
