<p>
Sharing some naming conventions is a key point to make it possible for a team to efficiently collaborate. 
This rule allows to check that all function names match a provided regular expression.
</p>

<h2>Noncompliant Code Example</h2>

<p>With default provided regular expression: ^[a-z_][a-z0-9_]{2,30}$</p>

<pre>
void MyFunction(int a, int b) {
	...
}
</pre>

<h2>Compliant Solution</h2>

<pre>
void my_function(int a, int b) {
	...
}
</pre>
